sympy/solvers/ode/tests/test_single.py

```diff
@@ -640,8 +467,213 @@ def _get_examples_ode_sol_euler_var_para():
+        'sol': [], #currently dsolve doesn't return any solution for this example
+        'XFAIL': ['factorable']
+    },
+    }
+    }
+
+def _get_examples_ode_sol_nth_algebraic():
+    M, m, r, t = symbols('M m r t')
+    phi = Function('phi')
+    # This one needs a substitution f' = g.
+    # 'algeb_12': {
+    #     'eq': -exp(x) + (x*Derivative(f(x), (x, 2)) + Derivative(f(x), x))/x,
+    #     'sol': [Eq(f(x), C1 + C2*log(x) + exp(x) - Ei(x))],
+    # },
+    return {
+            'hint': "nth_algebraic",
+            'func': f(x),
+            'examples':{
+    'algeb_01': {
+        'eq': f(x) * f(x).diff(x) * f(x).diff(x, x) * (f(x) - 1) * (f(x).diff(x) - x),
+        'sol': [Eq(f(x), C1 + x**2/2), Eq(f(x), C1 + C2*x)]
+    },
+
+    'algeb_02': {
+        'eq': f(x) * f(x).diff(x) * f(x).diff(x, x) * (f(x) - 1),
+        'sol': [Eq(f(x), C1 + C2*x)]
+    },
+
+    'algeb_03': {
+        'eq': f(x) * f(x).diff(x) * f(x).diff(x, x),
+        'sol': [Eq(f(x), C1 + C2*x)]
+    },
+
+    'algeb_04': {
+        'eq': Eq(-M * phi(t).diff(t),
+         Rational(3, 2) * m * r**2 * phi(t).diff(t) * phi(t).diff(t,t)),
+        'sol': [Eq(phi(t), C1), Eq(phi(t), C1 + C2*t - M*t**2/(3*m*r**2))],
+        'func': phi(t)
+    },
+
+    'algeb_05': {
+        'eq': (1 - sin(f(x))) * f(x).diff(x),
+        'sol': [Eq(f(x), C1)],
+        'XFAIL': ['separable']  #It raised exception.
+    },
+
+    'algeb_06': {
+        'eq': (diff(f(x)) - x)*(diff(f(x)) + x),
+        'sol': [Eq(f(x), C1 - x**2/2), Eq(f(x), C1 + x**2/2)]
+    },
+
+    'algeb_07': {
+        'eq': Eq(Derivative(f(x), x), Derivative(g(x), x)),
+        'sol': [Eq(f(x), C1 + g(x))],
+    },
+
+    'algeb_08': {
+        'eq': f(x).diff(x) - C1,   #this example is from issue 15999
+        'sol': [Eq(f(x), C1*x + C2)],
+    },
+
+    'algeb_09': {
+        'eq': f(x)*f(x).diff(x),
+        'sol': [Eq(f(x), C1)],
+    },
+
+    'algeb_10': {
+        'eq': (diff(f(x)) - x)*(diff(f(x)) + x),
+        'sol': [Eq(f(x), C1 - x**2/2), Eq(f(x), C1 + x**2/2)],
+    },
+
+    'algeb_11': {
+        'eq': f(x) + f(x)*f(x).diff(x),
+        'sol': [Eq(f(x), 0), Eq(f(x), C1 - x)],
+        'XFAIL': ['separable', '1st_exact', '1st_linear', 'Bernoulli', '1st_homogeneous_coeff_best',
+         '1st_homogeneous_coeff_subs_indep_div_dep', '1st_homogeneous_coeff_subs_dep_div_indep',
+         'lie_group', 'nth_linear_constant_coeff_undetermined_coefficients',
+         'nth_linear_euler_eq_nonhomogeneous_undetermined_coefficients',
+         'nth_linear_constant_coeff_variation_of_parameters',
+         'nth_linear_euler_eq_nonhomogeneous_variation_of_parameters']
+         #nth_linear_constant_coeff_undetermined_coefficients raises exception rest all of them misses a solution.
+    },
+
+    'algeb_12': {
+        'eq': Derivative(x*f(x), x, x, x),
+        'sol': [Eq(f(x), (C1 + C2*x + C3*x**2) / x)],
+        'XFAIL': ['nth_algebraic']  # It passes only when prep=False is set in dsolve.
```

Comment:
This example is correctly solved by `nth_algebraic` when you don't specify the hint:
```julia
In [6]: eq = Derivative(x*f(x), x, x, x)                                                                                          

In [7]: classify_ode(eq)                                                                                                          
Out[7]: 
('nth_algebraic',
 'nth_linear_euler_eq_homogeneous',
 'nth_order_reducible',
 'nth_algebraic_Integral')

In [8]: sol = dsolve(eq)                                                                                                          

In [9]: sol                                                                                                                       
Out[9]: 
            C₂       
f(x) = C₁ + ── + C₃⋅x
            x        

In [10]: checkodesol(eq, sol)                                                                                                     
Out[10]: (True, 0)

In [11]: dsolve(eq, f(x), hint='nth_algebraic')                                                                                   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-4a720b78911c> in <module>
----> 1 dsolve(eq, f(x), hint='nth_algebraic')

~/current/sympy/sympy/sympy/solvers/ode/ode.py in dsolve(eq, func, hint, simplify, ics, xi, eta, x0, n, **kwargs)
    616 
    617         # See the docstring of _desolve for more details.
--> 618         hints = _desolve(eq, func=func,
    619             hint=hint, simplify=True, xi=xi, eta=eta, type='ode', ics=ics,
    620             x0=x0, n=n, **kwargs)

~/current/sympy/sympy/sympy/solvers/deutils.py in _desolve(eq, func, hint, ics, simplify, **kwargs)
    269         raise ValueError("Hint not recognized: " + hint)
    270     elif hint not in hints:
--> 271         raise ValueError(string + str(eq) + " does not match hint " + hint)
    272     else:
    273         # Key added to identify the hint needed to solve the equation

ValueError: ODE x*Derivative(f(x), (x, 3)) + 3*Derivative(f(x), (x, 2)) does not match hint nth_algebraic
```
Maybe the preparation of the equation is different if the hint is specified or if it is not...

I guess that means that:

1. The `test_all_hints` function should also test what happens if no hint is specified.
2. There is something to be fixed in dsolve to make sure that the equation is prepared in the same way whether a hint is given or not.

Reply:
> 1. The `test_all_hints` functions should also test what happens if no hint is specified.

can you explain in more detail what is expected from this? like what things i need to test exactly?

Reply:
The test functions should also test what happens when calling `dsolve(eq, func)` to see if the ODE fails with the default hint. At the moment you only test with `dsolve(eq, func, hint=our_hint)`.

In this case it makes it look like `dsolve` can not solve this example with the `nth_algebraic` hint when actually it can but only if the hint is not provided in the call to dsolve.